### ISSUE TITLE
Error converting .all files v0.8.8

### ISSUE THREAD

**Issue by davesteps** [noaa-ocs-hydrography/kluster#79]
Hi, I'm getting the following error when trying to import/convert several .all files: 

```
Using existing local cluster client...
<Client: 'tcp://127.0.0.1:55569' processes=8 threads=8, memory=63.87 GiB>
****Running Kongsberg .all converter****
1 file(s), Using 1 chunk(s) in parallel
[                                        ] | 0% Completed |  1.4s
Error running action multibeam
Traceback (most recent call last):
  File "HSTB\kluster\gui\kluster_worker.py", line 38, in run
  File "HSTB\kluster\fqpr_actions.py", line 270, in execute_action
  File "HSTB\kluster\fqpr_actions.py", line 52, in execute
  File "HSTB\kluster\fqpr_convenience.py", line 112, in convert_multibeam
  File "HSTB\kluster\fqpr_generation.py", line 532, in read_from_source
  File "HSTB\kluster\xarray_conversion.py", line 1000, in read
  File "HSTB\kluster\xarray_conversion.py", line 1552, in batch_read
  File "HSTB\kluster\xarray_conversion.py", line 1314, in _batch_read_sequential
  File "distributed\client.py", line 1946, in gather
    return self.sync(
  File "distributed\utils.py", line 310, in sync
    return sync(
  File "distributed\utils.py", line 364, in sync
    raise exc.with_traceback(tb)
  File "distributed\utils.py", line 349, in f
    result[0] = yield future
  File "tornado\gen.py", line 762, in run
  File "distributed\client.py", line 1811, in _gather
    raise exception.with_traceback(traceback)
  File "HSTB\kluster\xarray_conversion.py", line 110, in _run_sequential_read
  File "HSTB\kluster\fqpr_drivers.py", line 173, in sequential_read_multibeam
  File "HSTB\drivers\par3.py", line 879, in sequential_read_records
  File "HSTB\drivers\par3.py", line 801, in _finalize_records
IndexError: index 1681 is out of bounds for axis 0 with size 1681
```
OS: WIndows 10
Version: 0.8.8 (same error with v0.8.4)

**Comment by ericgyounkin** [comment]
Are you able to process the file that I have in the repo?

https://github.com/noaa-ocs-hydrography/kluster/blob/master/tests/resources/0009_20170523_181119_FA2806.all

**Comment by davesteps** [comment]
> Are you able to process the file that I have in the repo?

Yes that has imported no problem

**Comment by davesteps** [comment]
Thanks I'll find out what the restrictions on this data are and whether I'm permitted to share it 

FYI I am getting a slightly different error on one of the other files:

```console
****Running Kongsberg .all converter****
1 file(s), Using 1 chunk(s) in parallel
[####################                    ] | 50% Completed |  0.4sdistributed.worker - WARNING - Compute Failed
Function:  _sequential_to_xarray
args:      ({'attitude': {'time': array([1.41769446e+09, 1.41769446e+09, 1.41769446e+09, ...,
       1.41769495e+09, 1.41769495e+09, 1.41769495e+09]), 'roll': array([1.14     , 1.15     , 1.15     , ..., 1.55     , 1.56     ,
       1.5699999], dtype=float32), 'pitch': array([-0.26, -0.26, -0.26, ...,  0.12,  0.11,  0.11], dtype=float32), 'heave': array([0.48, 0.47, 0.47, ..., 0.55, 0.55, 0.55], dtype=float32), 'heading': array([59.359997, 59.359997, 59.359997, ..., 53.449997, 53.449997,
       53.449997], dtype=float32)}, 'installation_params': {'time': array([1.41769446e+09]), 'serial_one': array([220], dtype=uint16), 'serial_two': array([0], dtype=uint16), 'installation_settings': array([{'waterline_vertical_location': '1.660', 'system_main_head_serial_number': '220', 'system_transducer_configuration': '0', 'transducer_1_along_location': '15.518', 'transducer_1_athwart_location': '-0.157', 'transducer_1_vertical_location': '9.626', 'transducer_1_heading_angle': '359.910', 'transducer_1_pitch_a
kwargs:    {}
Exception: 'ValueError("conflicting sizes for dimension \'time\': length 4911 on the data but length 5893 on coordinate \'time\'")'
Traceback (most recent call last):
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3251, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-57ad40619c65>", line 1, in <module>
    fq = convert_multibeam(r"/home/david/kluster/data/CEFAS_ISLES_OF_SCILLY_22-06-15_2020"
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/kluster/fqpr_convenience.py", line 112, in convert_multibeam
    fqpr_inst.read_from_source(build_offsets=False)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/kluster/fqpr_generation.py", line 532, in read_from_source
    self.multibeam.read(build_offsets=build_offsets)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/kluster/xarray_conversion.py", line 1000, in read
    final_pths = self.batch_read(self.filtype)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/kluster/xarray_conversion.py", line 1563, in batch_read
    input_xarrs = self._batch_read_sort_futures_by_time(input_xarrs)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/kluster/xarray_conversion.py", line 1341, in _batch_read_sort_futures_by_time
    mintims = self.client.gather(self.client.map(_return_xarray_mintime, input_xarrs))
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/distributed/client.py", line 1946, in gather
    return self.sync(
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/distributed/utils.py", line 310, in sync
    return sync(
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/distributed/utils.py", line 364, in sync
    raise exc.with_traceback(tb)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/distributed/utils.py", line 349, in f
    result[0] = yield future
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/distributed/client.py", line 1811, in _gather
    raise exception.with_traceback(traceback)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/kluster/xarray_conversion.py", line 307, in _sequential_to_xarray
    recs_to_merge[r][ky] = xr.DataArray(np.float32(rec[r][ky]), coords=[rec[r]['time']], dims=['time'])
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/xarray/core/dataarray.py", line 406, in __init__
    coords, dims = _infer_coords_and_dims(data.shape, coords, dims)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/xarray/core/dataarray.py", line 154, in _infer_coords_and_dims
    raise ValueError(
ValueError: conflicting sizes for dimension 'time': length 4911 on the data but length 5893 on coordinate 'time'

```

**Comment by ericgyounkin** [comment]
OK, great.  I have about 20 or so test datasets that I use for validation, would be good to get these files to determine what the issue might be.

I don't need many, maybe just one for each different error you are seeing.

Would you be able to tell me the sonar model number /  approximate year of the survey?  Just out of curiosity.

**Comment by monocilindro** [comment]
Is there any docker image/file of Kluster, @ericgyounkin? That could work in Ubuntu WSL2 and get the graphic user interface running in GWSL. It can be done in other software meant for Linux like MBsystem so you can run it in Windows without using any virtual machine software.

**Comment by ericgyounkin** [comment]
I have one in the latest branch that I haven't merged just yet.

[https://github.com/noaa-ocs-hydrography/kluster/blob/kluster_0_8_9/dockerfile](https://github.com/noaa-ocs-hydrography/kluster/blob/kluster_0_8_9/dockerfile)

I believe we'd need to uncomment that second to last line to actually complete the install, but you get the idea.  I toyed with it yesterday, it seemed to work, minus the graphical issues I was having that prevented the GUI from working.

**Comment by ericgyounkin** [comment]
Hi @monocilindro, can we take this into a separate issue?  I'd like to only address the .all processing here.

Would you be able to generate a new issue?

**Comment by monocilindro** [comment]
Yes, sure. My bad. It already done

**Comment by davesteps** [comment]
> Would you be able to tell me the sonar model number / approximate year of the survey? Just out of curiosity.

Those were an EM2040 collected in 2020. I also have some collected with a EM2040D to test.

**Comment by ericgyounkin** [comment]
@davesteps thanks.  Please let me know if you get permission to share one of these files.  I'd really like to take a look and resolve this issue!

**Comment by ericgyounkin** [comment]
@davesteps I just released a new version (v0.8.9) of Kluster that has a tool that might help here.  If you have time, you can try it by doing the following

- update to latest Kluster.  Should also update hstb.drivers

```
(kluster_test) C:\>pip install git+https://github.com/noaa-ocs-hydrography/kluster.git#egg=hstb.kluster
Collecting hstb.kluster
...
Successfully installed hstb.drivers-0.2.2 hstb.kluster-0.8.9
```

- open Kluster gui

`(kluster_test) C:\>python -m HSTB.kluster`

- Run Tools - File Analyzer on a small .all file that is not working in Kluster.  The kluster_read_test should print out some information about the file that we can use to figure out what is going on.  I'm not sure if you can share that info either, but at least it is more of a summary of the file.

![image](https://user-images.githubusercontent.com/58827984/154725861-4cd7948e-9af7-45e7-a757-36431af4ecb5.png)

**Comment by davesteps** [comment]
Thanks @ericgyounkin  I have run the read test and you can see the output below:

```
File: /home/david/kluster/data/test.all
File size: 17676586
Has Data110 record: False
Has older rangeangle datagram: False
===== 2022.02.18 19:02:04 =====
Traceback (most recent call last):
  File "/home/david/kluster/HSTB/kluster/gui/dialog_fileanalyzer.py", line 112, in run_function
    kluster_read_test(self.fileobject)
  File "/home/david/kluster/HSTB/kluster/fqpr_drivers.py", line 202, in kluster_read_test
    par3.kluster_read_test(file_object, byte_count=-1)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/drivers/par3.py", line 5793, in kluster_read_test
    data = file_object.sequential_read_records(first_installation_rec=False)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/drivers/par3.py", line 879, in sequential_read_records
    recs_to_read = self._finalize_records(recs_to_read, recs_count, sonarmodelnumber)
  File "/home/david/miniconda3/envs/kluster_test/lib/python3.8/site-packages/HSTB/drivers/par3.py", line 801, in _finalize_records
    dset[var] = dset[var][index]
IndexError: index 1681 is out of bounds for axis 0 with size 1681
```
The `read_first_fifty_records` test seems to return much more info here is the first few:

```
*****************************************************
Bytes : 941
Start : 2
Type : 73
Model : 2040
Date : 20141204
Time : 39318868
None
SurveyLine# : 0
Serial# : 220
Serial#2 : 0
1pps_clock_sync : falling_edge
active_attitude_velocity : attvel_not_used
active_heading_sensor : motion_1_com2
active_heave_sensor : com2
active_position_system_number : position_1
active_roll_pitch_sensor : com2
attitude_velocity_sensor_1_address : 3001
attitude_velocity_sensor_1_port : 0
attitude_velocity_sensor_2_address : 3000
attitude_velocity_sensor_2_port : 0
bsp_software_version : 1.2.0 130429
clock_offset : 0
clock_source : active_pos
datagram_format_version : 3.1.6 130104
dds_software_version : 3.5.9 130926
ethernet_2_address : 192.168.1.1
ethernet_2_network_mask : 255.255.0.0
gyrocompass_heading_offset : 0.000
motion_sensor_1_along_location : 0.000
motion_sensor_1_athwart_location : 0.000
motion_sensor_1_heading_angle : 0.000
motion_sensor_1_pitch_angle : 0.000
motion_sensor_1_roll_angle : 0.200
motion_sensor_1_roll_ref_plane : RP
motion_sensor_1_time_delay : 0
motion_sensor_1_vertical_location : 0.000
motion_sensor_2_along_location : 0.000
motion_sensor_2_athwart_location : 0.000
motion_sensor_2_heading_angle : 0.000
motion_sensor_2_pitch_angle : 0.000
motion_sensor_2_roll_angle : 0.000
motion_sensor_2_roll_ref_plane : RP
motion_sensor_2_time_delay : 0
motion_sensor_2_vertical_location : 0.000
operator_station_software_version : SIS 4.1.3
position_1_along_location : -3.533
position_1_athwart_location : -0.716
position_1_datum : WGS84
position_1_motion_compensation : passive
position_1_quality_check : 1
position_1_time_delay : 0.000
position_1_time_stamp : position_datagram
position_1_vertical_location : -21.716
position_2_along_location : 0.000
position_2_athwart_location : 0.000
position_2_datum : WGS84
position_2_motion_compensation : active
position_2_quality_check : 1
position_2_time_delay : 0.000
position_2_time_stamp : system_time
position_2_vertical_location : 0.000
position_3_along_location : 0.000
position_3_athwart_location : 0.000
position_3_datum : WGS84
position_3_motion_compensation : active
position_3_quality_check : 1
position_3_serial_or_ethernet : 1
position_3_time_delay : 0.000
position_3_time_stamp : system_time
position_3_vertical_location : 0.000
pressure_sensor_along_location : 0.000
pressure_sensor_athwart_location : 0.000
pressure_sensor_heave : not_depth_sensor
pressure_sensor_offset : 0.000000
pressure_sensor_scale_factor : 1.000000
pressure_sensor_time_delay : 0
pressure_sensor_vertical_location : 0.000
processing_unit_software_version : 1.3.3 140121
raw_file_name : test.all
receiver_software_version : 1.02 121112
roll_scaling_factor : 1.000
rx_array_size : 1
rx_no1_serial_number : 158
sonar_head_1_gain_offset : 0.00
sonar_model_number : em2040
survey_identifier : test
system_main_head_serial_number : 220
system_transducer_configuration : 0
transducer_1_along_location : 15.518
transducer_1_athwart_location : -0.157
transducer_1_heading_angle : 359.910
transducer_1_pitch_angle : -0.060
transducer_1_roll_angle : 0.800
transducer_1_vertical_location : 9.626
transducer_2_along_location : 15.518
transducer_2_athwart_location : 0.150
transducer_2_heading_angle : 0.230
transducer_2_pitch_angle : -0.290
transducer_2_roll_angle : 0.230
transducer_2_vertical_location : 9.614
transmitter_software_version : 1.07 140123
tx_array_size : 0
tx_serial_number : 110
waterline_vertical_location : 1.660

*****************************************************
Bytes : 52
Start : 2
Type : 82
Model : 2040
Date : 20141204
Time : 39318868
None
Counter : 25864
SystemSerial# : 220
OperatorStationStatus : 00000000
ProcessingUnitStatus : 00000000
BSPStatus : 00000000
SonarHeadOrTransceiverStatus : 00000000
Mode : 00000001
FilterID : 00001110
MinDepth : 5
MaxDepth : 150
AbsorptionCoefficent : 79.31
TransmitPulseLength : 0.0002
TransmitBeamWidth : 0.5
TransmitPower : 0
ReceiveBeamWidth : 1.0
ReceiveBandWidth50Hz : 255
ReceiverFixedGain : 5
TVGlawCrossoverAngle : 10
SourceOfSoundSpeed : 00000000
MaxPortSwathWidth : 90
BeamSpacing : 00000011
MaxPortCoverage : 65
YawAndPitchStabilization : 10000100
MaxStarboardCoverage : 65
MaxStarboardSwathWidth : 90
TransmitAlongTilt : 0.0
HiLoFrequencyAbsorptionCoeffRatio : 00010000

*****************************************************
Bytes : 52
Start : 2
Type : 82
Model : 2040
Date : 20141204
Time : 39318868
None
Counter : 25864
SystemSerial# : 220
OperatorStationStatus : 00000000
ProcessingUnitStatus : 00000000
BSPStatus : 00000000
SonarHeadOrTransceiverStatus : 00000000
Mode : 00000001
FilterID : 00001110
MinDepth : 5
MaxDepth : 150
AbsorptionCoefficent : 79.31
TransmitPulseLength : 0.0002
TransmitBeamWidth : 0.5
TransmitPower : 0
ReceiveBeamWidth : 1.0
ReceiveBandWidth50Hz : 255
ReceiverFixedGain : 5
TVGlawCrossoverAngle : 10
SourceOfSoundSpeed : 00000000
MaxPortSwathWidth : 90
BeamSpacing : 00000011
MaxPortCoverage : 65
YawAndPitchStabilization : 10000100
MaxStarboardCoverage : 65
MaxStarboardSwathWidth : 90
TransmitAlongTilt : 0.0
HiLoFrequencyAbsorptionCoeffRatio : 00010000

*****************************************************
Bytes : 792
Start : 2
Type : 85
Model : 2040
Date : 20141204
Time : 39318868
None
ProfileCounter : 1273
SystemSerial# : 220
Date : 20141204
Time : 34440.0
NumEntries : 95
DepthResolution : 1

POSIXtime : 1417690518.868

Depth : [0.00e+00 1.60e+00 1.80e+00 2.90e+00 4.60e+00 5.10e+00 5.80e+00 6.50e+00
 7.30e+00 8.30e+00 9.40e+00 1.06e+01 1.12e+01 1.16e+01 1.27e+01 1.39e+01
 1.47e+01 1.54e+01 1.64e+01 1.75e+01 1.81e+01 1.89e+01 1.98e+01 2.07e+01
 2.16e+01 2.26e+01 2.38e+01 2.43e+01 2.48e+01 2.61e+01 2.71e+01 2.75e+01
 2.85e+01 2.97e+01 3.07e+01 3.12e+01 3.18e+01 3.29e+01 3.37e+01 3.48e+01
 3.57e+01 3.63e+01 3.75e+01 3.83e+01 3.88e+01 3.96e+01 4.06e+01 4.16e+01
 4.28e+01 4.35e+01 4.41e+01 4.48e+01 4.57e+01 4.66e+01 4.79e+01 4.87e+01
 4.89e+01 4.96e+01 5.10e+01 5.20e+01 5.24e+01 5.31e+01 5.42e+01 5.49e+01
 5.53e+01 5.69e+01 5.81e+01 5.85e+01 5.88e+01 5.95e+01 6.10e+01 6.20e+01
 6.28e+01 6.37e+01 6.41e+01 6.46e+01 6.61e+01 6.76e+01 6.83e+01 6.88e+01
 6.98e+01 7.07e+01 7.11e+01 7.23e+01 7.32e+01 7.33e+01 5.00e+02 7.00e+02
 1.00e+03 1.50e+03 2.50e+03 4.00e+03 6.00e+03 8.50e+03 1.20e+04]
SoundSpeed : [1479.     1500.1    1500.1    1500.1    1500.1    1500.1    1500.2001
 1500.2001 1500.2001 1500.2001 1500.2001 1500.2001 1500.3    1500.3
 1500.3    1500.3    1500.3    1500.3    1500.3    1500.4    1500.4
 1500.4    1500.4    1500.4    1500.4    1500.4    1500.5    1500.5
 1500.5    1500.5    1500.5    1500.6    1500.6    1500.6    1500.6
 1500.6    1500.6    1500.6    1500.7001 1500.7001 1500.7001 1500.7001
 1500.7001 1500.7001 1500.8    1500.8    1500.8    1500.8    1500.8
 1500.8    1500.8    1500.8    1500.9    1500.9    1500.9    1500.9
 1500.9    1500.9    1501.     1501.     1501.     1501.     1501.
 1501.     1501.     1501.1    1501.1    1501.1    1501.1    1501.1
 1501.1    1501.1    1501.2001 1501.2001 1501.2001 1501.2001 1501.2001
 1501.2001 1501.3    1501.3    1501.3    1501.3    1501.3    1501.3
 1501.3    1501.3    1498.     1486.2001 1483.     1487.2001 1500.
 1524.1    1558.2001 1602.8    1669.    ]

*****************************************************
Bytes : 6508
```

**Comment by ericgyounkin** [comment]
@davesteps OK, well its good to know that it is in the .all driver exclusively.  That error means that either navigation or attitude have a variable that is not the same length as the concatenated time array.  Which shouldn't happen.  I see that this is an older-ish sonar, I wonder if this is a flavor of the .all format that I haven't seen.  There could be something here with decoding the gga/ggk nav or something.  I guess I would need to see a file to figure this out unfortunately.

**Comment by davesteps** [comment]
Thanks for the pointers, I don't think I'm going to get permission to send you this data. We will continue to see if we can debug our end. Have another file which is giving us errors when georeferencing, will open a separate issue.

**Comment by ericgyounkin** [comment]
@davesteps Understood.  I will try and add to the kluster_read_test to see if I can figure out the issue without getting the dataset.

**Comment by ericgyounkin** [comment]
Regarding the original error message you shared with me in this issue.  I have a fix of sorts in the drivers repository.  I think it will at least allow the kluster_read_test to finish.  I think it must be related to how we are using the Data65 Attitude datagram, where you somehow have some data where the arrays have different lengths.  

Could you try updating the drivers module and re-running the kluster_read_test for the file you used in the first message of this issue?  You can update using:

`pip install git+https://github.com/noaa-ocs-hydrography/drivers.git#egg=hstb.drivers`

If it works, it should issue some debug messages that we can use to figure out the cause of the problem

**Comment by davesteps** [comment]
Thanks @ericgyounkin, it is now able to read and geo-reference the data. A summary of the `kluster_read_test` is below:

```
File: <omitted>
File size: 17676586
Has Data110 record: False
Has older rangeangle datagram: False
par3: Unable to read 80: None - Altitude
par3: Unable to read 80: None - Altitude
par3: Unable to read 80: None - Altitude
...
par3: Unable to read 80: None - Altitude
par3: Unable to read 80: None - Altitude
par3: Unable to read 80: None - Altitude
par3: Unable to read 80: None - Altitude
Warning: variable altitude has a length of 1681, where time has a length of 2017, filling altitude by repeating last element 336 times
time (16806,) Found NaN values: 0 Found zero values: 0
[1.41769052e+09 1.41769052e+09 1.41769052e+09 ... 1.41769069e+09
 1.41769069e+09 1.41769069e+09]
roll (16806,) Found NaN values: 0 Found zero values: 27[ 1.67  1.67  1.67 ... -0.28 -0.28 -0.28]
pitch (16806,) Found NaN values: 0 Found zero values: 65
```

**Comment by ericgyounkin** [comment]
Thats great news, so some of the file had no stored navigation message.  It looks for a string header like "INGGA" or "INGGK", but it must not find it in some instances here.  That is strange.  I only use that part of the raw navigation blob for altitude, which is why altitude is specifically called out here.

So I can just set it so that a failure to read altitude results in parts of the final altitude record being NaN, which mirrors when altitude is unavailable in the message when the message is properly decoded.  

Resolved in:
https://github.com/noaa-ocs-hydrography/drivers/commit/2c2bf4de427616c97389b9a71f3bb9d50c9ea0a2
